pinot-tools/src/main/java/org/apache/pinot/tools/data/generator/StringGenerator.java

```diff
@@ -56,11 +65,18 @@ public void init() {
 
   @Override
   public Object next() {
+    if (numberOfValuesPerEntry == 1) {
+      return getNextString();
+    }
+    return MultiValueGeneratorHelper.generateMultiValueEntries(numberOfValuesPerEntry, rand, this::getNextString);
+  }
+
+  private String getNextString() {
     return vals.get(rand.nextInt(cardinality));
```

Comment:
Vals is a hashmap. This is where we are going to run into issues with generation based approach where the characteristics file specifies a very high cardinality for column(s). The HashMap will likely hit OOM

Reply:
I'm not sure why we need to generate millions of records. I assume something below 500K records should suffice and my assumption is that we won't get into trouble with that number of rows. 
BTW I have tested it on my mac with 10M rows with high cardinality on different columns in order of millions and no issue observed.

Reply:
It is not unusual to have million rows in a completed segment.

Reply:
That's correct, but even with completed segments, `MemoryEstimator` tries to extrapolate the actual memory size based on secondsToConsume (derived from numHours input) parameter:
```java
int totalDocs = (int) (((double) secondsToConsume / _sampleSegmentConsumedSeconds) * _totalDocsInSampleSegment);
long memoryForConsumingSegmentPerPartition = getMemoryForConsumingSegmentPerPartition(statsFile, totalDocs);
```
My point we have only one completed segment (given or generated) and that should be enough to for estimating memory for different consumption window sizes.